### PR TITLE
NO-ISSUE: Add OKD support

### LINKED ISSUE #2580: Make it work with OKD
The docs should contain a hint of how AI can be used to deploy OKD instead of OCP

### PR DESCRIPTION
**Pull request by vrutkovs**
This PR adds assisted-installer support for OKD clusters. Most notable difference between OCP and OKD install is that OKD starts with Fedora CoreOS instead of RHCOS. FCOS doesn't have kubelet/crio/mcd (but has podman), so in order to make bootstrap work on live CD we do the following:
* pull an image with RPMs (`OKD_RPMS_IMAGE` env var)
* create an overlayFS for `/usr` (as its read-only on rpm-ostree systems)
* extract RPMs from the image and use `rpm` to install them
* expand `/var` and `/run` partitions to 6 GB (in FCOS these partitions are 3.2 GB for 16 GB VM which is insufficient for bootstrap to complete)
* hold off agent.service until RPMs are installed
* add fake `systemd-journal-gatewayd.socket` (required from `mcd --once-from` to complete, as FCOS doesn't have journal gateway preinstalled)

This PR also makes easy to deploy OKD-specific version of AI via podman: use `make deploy-onprem OKD=true` to deploy service using `okd-configmap.yml`

TODO:
* [x] Update `generate.sh` to have `okd-configmap.yml` autoupdated when `configmap.yml` is being generated
       Fixed in 3327bcd1f8ab3a8a83b1d84a0161f3804510374e
* [x] Allow `agent.service` to restart on failure. This would make `systemctl start agent` unnecessary.
       Possible unintended consequences?
       Fixed in b84acd808584a68d6478b69121bbe59dd76eaf70 - `Wants/After` is a better solution (agent.service already restarts on failure)

## List all the issues related to this PR

- [x] Fixes #2580 
- [x] Enhancement 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @romfreiman 
/cc @tsorya 
/cc @eranco74 

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: deploy/podman/okd-configmap.yml
Comment (omertuc) [inline deploy/podman/okd-configmap.yml:1]: A lot of the values in the original configmap are auto-generated https://github.com/openshift/assisted-service/blob/a9869850e72b002f4e9f89374da2bfd2e60dc43a/hack/generate.sh#L100-L103

Comment (omertuc) [reply deploy/podman/okd-configmap.yml:1]: We also have [automation](https://github.com/openshift-assisted/assisted-installer-deployment/blob/454edbd4fea43c99f6bffc816a3426248a63fa2c/tools/bump_ocp_releases.py#L1) in place to periodically create PRs which bump the z-stream version, you'll probably want to do the same for OKD too

Comment (vrutkovs) [reply deploy/podman/okd-configmap.yml:1]: Good idea, added a TODO

File: internal/ignition/ignition.go
Comment (omertuc) [inline internal/ignition/ignition.go:139]: Add a small sleep here, just in-case podman fails on something very early, we probably don't want to spin too quickly

File: internal/ignition/ignition.go
Comment (omertuc) [inline internal/ignition/ignition.go:161]: Isn't the agent already waiting on `/tmp/.okd-rpms-installed`? Why does it need a `start`?

Comment (vrutkovs) [reply internal/ignition/ignition.go:161]: `agent` is not set to autorestart on failure, so it needs to be kickstarted. I suppose we could update the service, but not sure about unintended effects of this

Comment (omertuc) [reply internal/ignition/ignition.go:161]: >autorestart on failure
What failure is it hitting? (assuming it's just waiting on your file to show up I wouldn't expect it to fail)

Comment (omertuc) [reply internal/ignition/ignition.go:161]: >I suppose we could update the service, but not sure about unintended effects of this

That sounds like a good idea, let me verify if this wasn't done on purpose for some good reason

Comment (omertuc) [reply internal/ignition/ignition.go:161]: Looks (f6403fb153d45bcb3204b61e28c410a6ec029fe2) like it was restarting early on, trying to find the commit where someone removed it

Comment (omertuc) [reply internal/ignition/ignition.go:161]: >agent is not set to autorestart on failure

Oh wait this doesn't seem true
It is set to auto restart on anything, not just failure https://github.com/openshift/assisted-service/blob/70cd21e2b64f0e28993182c29cae733242e359a9/internal/ignition/ignition.go#L147

Comment (vrutkovs) [reply internal/ignition/ignition.go:161]: It would be blocked by absence of `/tmp/.okd-rpms-installed` (which systemd treats as not started), but it won't autorestart

Comment (omertuc) [reply internal/ignition/ignition.go:161]: I'm confused, if it's blocked by the absence of okd-rpms-installed, shouldn't systemd be watching for the existence of this file and start the service when that file shows up?

Or is that not how it works?

Comment (vrutkovs) [reply internal/ignition/ignition.go:161]: >shouldn't systemd be watching for the existence of this file and start the service when that file shows up, automatically?

Not if you have Restart unset or disabled :/

Comment (omertuc) [reply internal/ignition/ignition.go:161]: But it is set 
image

Comment (vrutkovs) [reply internal/ignition/ignition.go:161]: Ah, interesting. I'll play with that some more then

**Review by mhrivnak — CHANGES_REQUESTED** [review]

File: deploy/podman/okd-configmap.yml
Comment (vrutkovs) [reply deploy/podman/okd-configmap.yml:31]: Later on we'll build this image as part of https://github.com/openshift/okd-machine-os and have it included in OKD payload and mirrored to quay.

File: deploy/podman/okd-configmap.yml
Comment (vrutkovs) [reply deploy/podman/okd-configmap.yml:2]: Updated README in 51e9e91ce34fa0df3ef95aaa49fd488678a0b356

File: internal/ignition/ignition.go
Comment (vrutkovs) [reply internal/ignition/ignition.go]: Reworked in b84acd808584a68d6478b69121bbe59dd76eaf70 - OKD installs would add `Wants/After` dropin for agent service

File: internal/ignition/ignition.go
Comment (vrutkovs) [reply internal/ignition/ignition.go:139]: Fixed in e91b88ece692a523f852439fa2c7fdbf22f88189

File: deploy/podman/okd-configmap.yml
Comment (vrutkovs) [reply deploy/podman/okd-configmap.yml:1]: Fixed in 3327bcd1f8ab3a8a83b1d84a0161f3804510374e

File: hack/generate.sh
Comment (omertuc) [inline hack/generate.sh:100]: Consider creating `assisted-service/data/default_okd_release_images.json` and `assisted-service/data/default_okd_os_images.json`, so the source of truth can be entirely in `data/`

Comment (vrutkovs) [reply hack/generate.sh:102]: Ah, good idea. Fixed in f277d54031fa945c805b0c7a5d8e1204b64b49de

**Comment by vrutkovs** [comment]
continuous-integration/jenkins/pr-merge won't pass due to invalid commit messages - I'll squash them in several smaller ones once we're happy with the code changes

File: hack/generate.sh
Comment (omertuc) [reply hack/generate.sh:102]: :ok_hand:

**Comment by omertuc** [comment]
Looks great as far as I'm concerned. @mhrivnak if you don't have any more concerns please /lgtm

/approve

**Comment by vrutkovs** [comment]
Reformatted commits to have `pr-merge` test pass

**Review by mhrivnak — CHANGES_REQUESTED** [review]

File: deploy/podman/okd-configmap.yml
Comment (LorbusChris) [reply deploy/podman/okd-configmap.yml:31]: Also somewhat related to this: https://github.com/openshift/os/issues/498
If that were implemented, it would re-align things nicely.

**Pull request merged**
